File: MonkeyJump/GameObjects/Monkey.m
//
//  Monkey.m
//  MonkeyJump
//
//  Created by Kauserali on 18/11/13.
//  Copyright (c) 2013 Raywenderlich. All rights reserved.
//

#import "Monkey.h"

#define kWalkAnimationKey @"walkAnimation"
#define kJumpAnimationKey @"jumpAnimation"

@implementation Monkey {
    SKAction *_walkAnimation, *_jumpAnimation;
}

+(instancetype)monkey {
    Monkey *monkey = [Monkey spriteNodeWithImageNamed:@"monkey_run1.png"];
    return monkey;
}

- (id)initWithTexture:(SKTexture *)texture color:(UIColor *)color size:(CGSize)size
{
    self = [super initWithTexture:texture color:color size:size];
    if (self) {
        self.gameElementType = kMonkey;
        self.lives = 3;
        [self initAnimations];
    }
    return self;
}

- (void)initAnimations
{
    _walkAnimation = [self loadPlistForAnimationName:@"walkAnim" andClassName:NSStringFromClass([self class])];
    _jumpAnimation = [self loadPlistForAnimationName:@"jumpAnim" andClassName:NSStringFromClass([self class])];
}

- (void)setState:(enum MonkeyState)state
{
    if (_state == state) {
        return;
    }
    _state = state;
    SKAction *action = nil;
    NSString *key;
    if (state == kWalking) {
        if ([self actionForKey:kJumpAnimationKey]) {
            [self removeActionForKey:kJumpAnimationKey];
        }
        action = [SKAction repeatActionForever:_walkAnimation];
        key = kWalkAnimationKey;
    } else if(state == kJumping) {
        if ([self actionForKey:kWalkAnimationKey]) {
            [self removeActionForKey:kWalkAnimationKey];
        }
        action = _jumpAnimation;
        key = kJumpAnimationKey;
    } else if(state == kDead) {
        [self removeAllActions];
        SKTextureAtlas *gameAtlas = [SKTextureAtlas atlasNamed:@"characteranimations"];
        [self setTexture:[gameAtlas textureNamed:@"monkey_dead.png"]];
    }
    if (action != nil) {
        [self runAction:action withKey:key];
    }
}
@end

File: MonkeyJump/GameObjects/Monkey.swift
//
//  Croc.swift
//  MonkeyJump
//
//  Original Objective-C Created by Kauserali on 18/11/13
//  Translated to Swift by Phil Eggel on 12/01/16
//  Copyright © 2016 PhilEagleDev.com. All rights reserved.
//

import SpriteKit

enum MonkeyState {
    case Idle, Walking, Jumping, Dead
}

class Monkey: GameElement {
    
    static let walkAnimationKey = "walkAnimation"
    static let jumpAnimationKey = "jumpAnimation"
    
    private var walkAnimation: SKAction!
    private var jumpAnimation: SKAction!
    
    var lives: Int
    var state: MonkeyState {
        didSet {
            if oldValue != state {
                selectCurrentAction()
            }
        }
    }
    
    private func selectCurrentAction() {
        
        switch state {
        case .Walking:
            if let _ = actionForKey(Monkey.jumpAnimationKey) {
                removeActionForKey(Monkey.jumpAnimationKey)
            }
            runAction(SKAction.repeatActionForever(walkAnimation), withKey: Monkey.walkAnimationKey)
        
        case .Jumping:
            if let _ = actionForKey(Monkey.walkAnimationKey) {
                removeActionForKey(Monkey.walkAnimationKey)
            }
            runAction(jumpAnimation, withKey: Monkey.jumpAnimationKey)
        
        case .Dead:
            removeAllActions()
            let gameAtlas = SKTextureAtlas(named: "characteranimations")
            texture = gameAtlas.textureNamed("monkey_dead.png")
        
        case .Idle:
            break
        }
    }
    
    override init(texture: SKTexture?, color: UIColor, size: CGSize) {
        self.lives = 3
        self.state = .Idle
        
        super.init(texture: texture, color: color, size: size)
        self.gameElementType = .Monkey
        self.walkAnimation = loadPlistForAnimationName("walkAnim", andClassName: String(self.dynamicType))
        self.jumpAnimation = loadPlistForAnimationName("jumpAnim", andClassName: String(self.dynamicType))
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    

}
